glide-core/src/ffi/ffi_client.rs

```diff
@@ -0,0 +1,472 @@
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, CreateClientError> {
+    let request = unsafe { create_connection_request_using_config(config) };
+    create_client_using_connection_request(request, success_callback, failure_callback)
+}
+
+fn create_client_using_protobuf_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, CreateClientError> {
+    let request = create_connection_request_using_bytes(connection_request_bytes)?;
+    create_client_using_connection_request(request, success_callback, failure_callback)
+}
+
+// TODO PR #1088
+fn create_client_using_connection_request(
+    connection_request: connection_request::ConnectionRequest,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, CreateClientError> {
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let redis_error = err.into();
+            CreateClientError {
+                message: error_message(&redis_error),
+                error_type: error_type(&redis_error),
+            }
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime
+        .block_on(GlideClient::new(connection_request))
+        .map_err(|err| CreateClientError {
+            message: err.to_string(),
+            error_type: RequestErrorType::Disconnect,
+        })?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+fn create_client_result_to_connection_response(result: Result<Client, CreateClientError>) -> ConnectionResponse {
+    match result {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            error_message: CString::into_raw(CString::new(err.message).unwrap()),
+            error_type: err.error_type,
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            error_message: std::ptr::null(),
+            error_type: RequestErrorType::Unspecified,
+        }
+    }
+}
+
+/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
+///
+/// # Safety
+///
+/// * `config` must be a valid [`ConnectionConfig`](ConnectionConfig) pointer. See the safety documentation of [`create_client_internal`](create_client_using_config_internal).
+#[allow(rustdoc::redundant_explicit_links)]
+#[allow(rustdoc::private_intra_doc_links)]
+#[no_mangle]
+pub unsafe extern "C" fn create_client_using_config(
+    config: *const ConnectionConfig,
```

Comment:
I recommend that you don't rely on structs in public APIs, since it requires a caller that can setup a struct - this capability isn't universally available. I know it's verbose, but it's much better to just take all of the arguments instead of wrapping them in a struct.
It also removes the need to manage the struct's lifetime around this call.

Reply:
not to do this in a unified way, instead implementing this for every language according to its capabilities. 🤷 
I agree its problematic, but it's better to face this problem ahead of time instead of creating an API and later understanding that it won't work.

Reply:
Okay, I think for now, we'll just have a separate lib.rs for each wrapper instead of taking this unified approach. If we find there are some parts that can reasonably be shared later on, we can move those to a shared library later.